File: android/ac-core/src/main/kotlin/com/microsoft/adaptivecards/core/hostconfig/HostConfig.kt
package com.microsoft.adaptivecards.core.hostconfig

import com.microsoft.adaptivecards.core.models.*
import kotlinx.serialization.Serializable

@Serializable
data class HostConfig(
    val fontFamily: String = "",
    val supportsInteractivity: Boolean = true,
    val imageBaseUrl: String = "",
    val spacing: SpacingConfig = SpacingConfig(),
    val separator: SeparatorConfig = SeparatorConfig(),
    val fontTypes: FontTypesConfig = FontTypesConfig(),
    val fontSizes: FontSizesConfig = FontSizesConfig(),
    val fontWeights: FontWeightsConfig = FontWeightsConfig(),
    val containerStyles: ContainerStylesConfig = ContainerStylesConfig(),
    val imageSizes: ImageSizesConfig = ImageSizesConfig(),
    val actions: ActionsConfig = ActionsConfig(),
    val adaptiveCard: AdaptiveCardConfig = AdaptiveCardConfig(),
    val imageSet: ImageSetConfig = ImageSetConfig(),
    val media: MediaConfig = MediaConfig(),
    val factSet: FactSetConfig = FactSetConfig(),
    val inputs: InputsConfig = InputsConfig(),
    val hostWidth: HostWidthConfig = HostWidthConfig(),
    val textBlock: TextBlockConfig = TextBlockConfig(),
    val textStyles: TextStylesConfig = TextStylesConfig(),
    val image: ImageConfig = ImageConfig(),
    val ratingLabel: RatingElementConfig = RatingElementConfig(),
    val ratingInput: RatingElementConfig = RatingElementConfig(),
    val table: TableConfig = TableConfig(),
    val compoundButton: CompoundButtonConfig = CompoundButtonConfig(),
    val lineHeights: LineHeightsConfig = LineHeightsConfig(),
    val borderWidth: Map<String, Int> = emptyMap(),
    val cornerRadius: CornerRadiusConfig = CornerRadiusConfig(),
    val badgeStyles: BadgeStylesConfig = BadgeStylesConfig(),
    val pageControl: PageControlConfig = PageControlConfig()
)

@Serializable
data class SpacingConfig(
    val extraSmall: Int = 4,
    val small: Int = 8,
    val default: Int = 12,
    val medium: Int = 16,
    val large: Int = 20,
    val extraLarge: Int = 24,
    val padding: Int = 12
)

@Serializable
data class SeparatorConfig(
    val lineThickness: Int = 1,
    val lineColor: String = "#EEEEEE"
)

@Serializable
data class FontTypesConfig(
    val default: FontTypeConfig = FontTypeConfig(),
    val monospace: FontTypeConfig = FontTypeConfig(fontFamily = "Courier New, Courier, monospace")
)

@Serializable
data class FontTypeConfig(
    val fontFamily: String = "Roboto, system-ui, sans-serif",
    val fontSizes: FontSizesConfig? = null,
    val fontWeights: FontWeightsConfig? = null
)

@Serializable
data class FontSizesConfig(
    val small: Int = 12,
    val default: Int = 14,
    val medium: Int = 14,
    val large: Int = 16,
    val extraLarge: Int = 20
)

/** Line height configuration matching the Figma type ramp */
@Serializable
data class LineHeightsConfig(
    val small: Int = 16,
    val default: Int = 18,
    val medium: Int = 18,
    val large: Int = 24,
    val extraLarge: Int = 24
)

@Serializable
data class FontWeightsConfig(
    val lighter: Int = 400,
    val default: Int = 400,
    val bolder: Int = 500
)

@Serializable
data class ContainerStylesConfig(
    val default: ContainerStyleConfig = ContainerStyleConfig(),
    val emphasis: ContainerStyleConfig = ContainerStyleConfig(backgroundColor = "#F1F1F1"),
    val good: ContainerStyleConfig = ContainerStyleConfig(backgroundColor = "#DFF6DD"),
    val attention: ContainerStyleConfig = ContainerStyleConfig(backgroundColor = "#FED9CC"),
    val warning: ContainerStyleConfig = ContainerStyleConfig(backgroundColor = "#FFF4CE"),
    val accent: ContainerStyleConfig = ContainerStyleConfig(backgroundColor = "#E8E8F7")
)

@Serializable
data class ContainerStyleConfig(
    val backgroundColor: String = "#FFFFFF",
    val borderColor: String = "#E0E0E0",
    val foregroundColors: ForegroundColorsConfig = ForegroundColorsConfig()
)

@Serializable
data class ForegroundColorsConfig(
    val default: ColorConfig = ColorConfig(default = "#212121", subtle = "#6E6E6E"),
    val dark: ColorConfig = ColorConfig(default = "#000000", subtle = "#212121"),
    val light: ColorConfig = ColorConfig(default = "#FFFFFF", subtle = "#F1F1F1"),
    val accent: ColorConfig = ColorConfig(default = "#6264A7", subtle = "#8B8CC7"),
    val good: ColorConfig = ColorConfig(default = "#237B4B", subtle = "#217346"),
    val warning: ColorConfig = ColorConfig(default = "#C50F1F", subtle = "#CC4A31"),
    val attention: ColorConfig = ColorConfig(default = "#C4314B", subtle = "#B24782")
)

@Serializable
data class ColorConfig(
    val default: String,
    val subtle: String,
    val highlightColors: HighlightColorConfig = HighlightColorConfig()
)

/** Highlight/selection colors for text within a color slot */
@Serializable
data class HighlightColorConfig(
    val default: String = "#FFFFFF00",
    val subtle: String = "#FFFFFFE0"
)

@Serializable
data class ImageSizesConfig(
    val small: Int = 32,
    val medium: Int = 52,
    val large: Int = 100
)

@Serializable
data class ActionsConfig(
    val maxActions: Int = 6,
    val spacing: Spacing = Spacing.Medium,
    val buttonSpacing: Int = 8,
    val showCard: ShowCardConfig = ShowCardConfig(),
    val actionsOrientation: String = "horizontal",
    val actionAlignment: String = "left",
    val iconPlacement: String = "leftOfTitle",
    val iconSize: Int = 24
)

@Serializable
data class ShowCardConfig(
    val actionMode: String = "inline",
    val style: ContainerStyle = ContainerStyle.Emphasis,
    val inlineTopMargin: Int = 16
)

@Serializable
data class AdaptiveCardConfig(
    val allowCustomStyle: Boolean = true
)

@Serializable
data class ImageSetConfig(
    val imageSize: ImageSize = ImageSize.Medium,
    val maxImageHeight: Int = 100
)

@Serializable
data class MediaConfig(
    val defaultPoster: String? = null,
    val playButton: String? = null,
    val allowInlinePlayback: Boolean = true
)

@Serializable
data class FactSetConfig(
    val title: FactSetTextConfig = FactSetTextConfig(weight = FontWeight.Bolder),
    val value: FactSetTextConfig = FactSetTextConfig(),
    val spacing: Int = 32
)

@Serializable
data class FactSetTextConfig(
    val size: FontSize = FontSize.Default,
    val color: Color = Color.Default,
    val isSubtle: Boolean = false,
    val weight: FontWeight = FontWeight.Default,
    val fontType: String = "Default",
    val wrap: Boolean = true,
    val maxWidth: Int = 0
)

@Serializable
data class InputsConfig(
    val label: InputLabelGroupConfig = InputLabelGroupConfig(),
    val errorMessage: InputErrorMessageConfig = InputErrorMessageConfig()
)

/** Label configuration for input groups (required vs optional) */
@Serializable
data class InputLabelGroupConfig(
    val inputSpacing: Spacing = Spacing.Default,
    val requiredInputs: InputLabelConfig = InputLabelConfig(),
    val optionalInputs: InputLabelConfig = InputLabelConfig()
)

@Serializable
data class InputLabelConfig(
    val color: Color = Color.Default,
    val isSubtle: Boolean = false,
    val size: FontSize = FontSize.Default,
    val suffix: String = "",
    val weight: FontWeight = FontWeight.Default
)

@Serializable
data class InputErrorMessageConfig(
    val color: Color = Color.Attention,
    val size: FontSize = FontSize.Small,
    val weight: FontWeight = FontWeight.Default
)

// MARK: - New configs ported from production

/** Responsive breakpoints for host width */
@Serializable
data class HostWidthConfig(
    val veryNarrow: Int = 216,
    val narrow: Int = 413,
    val standard: Int = 500
)

/** TextBlock-specific configuration */
@Serializable
data class TextBlockConfig(
    val headingLevel: Int = 2
)

/** Semantic text styles (heading, columnHeader, etc.) */
@Serializable
data class TextStylesConfig(
    val heading: TextStyleConfig = TextStyleConfig(),
    val columnHeader: TextStyleConfig = TextStyleConfig(weight = FontWeight.Bolder)
)

@Serializable
data class TextStyleConfig(
    val weight: FontWeight = FontWeight.Default,
    val size: FontSize = FontSize.Default,
    val isSubtle: Boolean = false,
    val color: Color = Color.Default,
    val fontType: String = "Default"
)

/** Default image configuration */
@Serializable
data class ImageConfig(
    val imageSize: ImageSize = ImageSize.Auto
)

/** Rating element (stars) configuration */
@Serializable
data class RatingElementConfig(
    val filledStar: RatingStarConfig = RatingStarConfig(),
    val emptyStar: RatingStarConfig = RatingStarConfig(),
    val ratingTextColor: String = "#000000",
    val countTextColor: String = "#000000"
)

@Serializable
data class RatingStarConfig(
    val marigoldColor: String = "#EAA300",
    val neutralColor: String = "#212121"
)

/** Table configuration */
@Serializable
data class TableConfig(
    val cellSpacing: Int = 8
)

/** Compound button configuration */
@Serializable
data class CompoundButtonConfig(
    val badge: BadgeConfig = BadgeConfig(),
    val borderColor: String = "#E1E1E1"
)

@Serializable
data class BadgeConfig(
    val backgroundColor: String = "#5B5FC7"
)

/** Corner radius configuration for containers and elements */
@Serializable
data class CornerRadiusConfig(
    val columnSet: Int = 4,
    val column: Int = 4,
    val container: Int = 4,
    val table: Int = 4,
    val image: Int = 4
)

/** Badge styles configuration from Figma spec */
@Serializable
data class BadgeStylesConfig(
    val default: BadgeStyleVariants = BadgeStyleVariants(),
    val accent: BadgeStyleVariants = BadgeStyleVariants(),
    val attention: BadgeStyleVariants = BadgeStyleVariants(),
    val good: BadgeStyleVariants = BadgeStyleVariants(),
    val informative: BadgeStyleVariants = BadgeStyleVariants(),
    val subtle: BadgeStyleVariants = BadgeStyleVariants(),
    val warning: BadgeStyleVariants = BadgeStyleVariants()
)

@Serializable
data class BadgeStyleVariants(
    val filled: BadgeStyleDef = BadgeStyleDef(),
    val tint: BadgeStyleDef = BadgeStyleDef()
)

@Serializable
data class BadgeStyleDef(
    val backgroundColor: String = "#212121",
    val strokeColor: String = "#212121",
    val textColor: String = "#FFFFFF"
)

/** Page control configuration */
@Serializable
data class PageControlConfig(
    val selectedTintColor: String = "#5B5FC7"
)

File: ios/Sources/ACCore/HostConfig/HostConfig.swift
import Foundation

public struct HostConfig: Codable {
    // MARK: - Core Properties
    public var fontFamily: String
    public var supportsInteractivity: Bool
    public var imageBaseUrl: String

    // MARK: - Existing Configs
    public var spacing: SpacingConfig
    public var separator: SeparatorConfig
    public var fontSizes: FontSizesConfig
    public var fontWeights: FontWeightsConfig
    public var fontTypes: FontTypesConfig
    public var containerStyles: ContainerStylesConfig
    public var imageSizes: ImageSizesConfig
    public var actions: ActionsConfig
    public var adaptiveCard: AdaptiveCardConfig
    public var imageSet: ImageSetConfig
    public var factSet: FactSetConfig

    // MARK: - New Configs (ported from production)
    public var media: MediaConfig
    public var inputs: InputsConfig
    public var hostWidth: HostWidthConfig
    public var textBlock: TextBlockConfig
    public var textStyles: TextStylesConfig
    public var image: ImageConfig
    public var ratingLabel: RatingElementConfig
    public var ratingInput: RatingElementConfig
    public var table: TableConfig
    public var compoundButton: CompoundButtonConfig
    public var borderWidth: [String: Int]
    public var cornerRadius: [String: Int]

    public init(
        fontFamily: String = "",
        supportsInteractivity: Bool = true,
        imageBaseUrl: String = "",
        spacing: SpacingConfig = SpacingConfig(),
        separator: SeparatorConfig = SeparatorConfig(),
        fontSizes: FontSizesConfig = FontSizesConfig(),
        fontWeights: FontWeightsConfig = FontWeightsConfig(),
        fontTypes: FontTypesConfig = FontTypesConfig(),
        containerStyles: ContainerStylesConfig = ContainerStylesConfig(),
        imageSizes: ImageSizesConfig = ImageSizesConfig(),
        actions: ActionsConfig = ActionsConfig(),
        adaptiveCard: AdaptiveCardConfig = AdaptiveCardConfig(),
        imageSet: ImageSetConfig = ImageSetConfig(),
        factSet: FactSetConfig = FactSetConfig(),
        media: MediaConfig = MediaConfig(),
        inputs: InputsConfig = InputsConfig(),
        hostWidth: HostWidthConfig = HostWidthConfig(),
        textBlock: TextBlockConfig = TextBlockConfig(),
        textStyles: TextStylesConfig = TextStylesConfig(),
        image: ImageConfig = ImageConfig(),
        ratingLabel: RatingElementConfig = RatingElementConfig(),
        ratingInput: RatingElementConfig = RatingElementConfig(),
        table: TableConfig = TableConfig(),
        compoundButton: CompoundButtonConfig = CompoundButtonConfig(),
        borderWidth: [String: Int] = [:],
        cornerRadius: [String: Int] = [:]
    ) {
        self.fontFamily = fontFamily
        self.supportsInteractivity = supportsInteractivity
        self.imageBaseUrl = imageBaseUrl
        self.spacing = spacing
        self.separator = separator
        self.fontSizes = fontSizes
        self.fontWeights = fontWeights
        self.fontTypes = fontTypes
        self.containerStyles = containerStyles
        self.imageSizes = imageSizes
        self.actions = actions
        self.adaptiveCard = adaptiveCard
        self.imageSet = imageSet
        self.factSet = factSet
        self.media = media
        self.inputs = inputs
        self.hostWidth = hostWidth
        self.textBlock = textBlock
        self.textStyles = textStyles
        self.image = image
        self.ratingLabel = ratingLabel
        self.ratingInput = ratingInput
        self.table = table
        self.compoundButton = compoundButton
        self.borderWidth = borderWidth
        self.cornerRadius = cornerRadius
    }

    // Custom decoder for backward compatibility - missing keys get defaults
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        fontFamily = try c.decodeIfPresent(String.self, forKey: .fontFamily) ?? ""
        supportsInteractivity = try c.decodeIfPresent(Bool.self, forKey: .supportsInteractivity) ?? true
        imageBaseUrl = try c.decodeIfPresent(String.self, forKey: .imageBaseUrl) ?? ""
        spacing = try c.decodeIfPresent(SpacingConfig.self, forKey: .spacing) ?? SpacingConfig()
        separator = try c.decodeIfPresent(SeparatorConfig.self, forKey: .separator) ?? SeparatorConfig()
        fontSizes = try c.decodeIfPresent(FontSizesConfig.self, forKey: .fontSizes) ?? FontSizesConfig()
        fontWeights = try c.decodeIfPresent(FontWeightsConfig.self, forKey: .fontWeights) ?? FontWeightsConfig()
        fontTypes = try c.decodeIfPresent(FontTypesConfig.self, forKey: .fontTypes) ?? FontTypesConfig()
        containerStyles = try c.decodeIfPresent(ContainerStylesConfig.self, forKey: .containerStyles) ?? ContainerStylesConfig()
        imageSizes = try c.decodeIfPresent(ImageSizesConfig.self, forKey: .imageSizes) ?? ImageSizesConfig()
        actions = try c.decodeIfPresent(ActionsConfig.self, forKey: .actions) ?? ActionsConfig()
        adaptiveCard = try c.decodeIfPresent(AdaptiveCardConfig.self, forKey: .adaptiveCard) ?? AdaptiveCardConfig()
        imageSet = try c.decodeIfPresent(ImageSetConfig.self, forKey: .imageSet) ?? ImageSetConfig()
        factSet = try c.decodeIfPresent(FactSetConfig.self, forKey: .factSet) ?? FactSetConfig()
        media = try c.decodeIfPresent(MediaConfig.self, forKey: .media) ?? MediaConfig()
        inputs = try c.decodeIfPresent(InputsConfig.self, forKey: .inputs) ?? InputsConfig()
        hostWidth = try c.decodeIfPresent(HostWidthConfig.self, forKey: .hostWidth) ?? HostWidthConfig()
        textBlock = try c.decodeIfPresent(TextBlockConfig.self, forKey: .textBlock) ?? TextBlockConfig()
        textStyles = try c.decodeIfPresent(TextStylesConfig.self, forKey: .textStyles) ?? TextStylesConfig()
        image = try c.decodeIfPresent(ImageConfig.self, forKey: .image) ?? ImageConfig()
        ratingLabel = try c.decodeIfPresent(RatingElementConfig.self, forKey: .ratingLabel) ?? RatingElementConfig()
        ratingInput = try c.decodeIfPresent(RatingElementConfig.self, forKey: .ratingInput) ?? RatingElementConfig()
        table = try c.decodeIfPresent(TableConfig.self, forKey: .table) ?? TableConfig()
        compoundButton = try c.decodeIfPresent(CompoundButtonConfig.self, forKey: .compoundButton) ?? CompoundButtonConfig()
        borderWidth = try c.decodeIfPresent([String: Int].self, forKey: .borderWidth) ?? [:]
        cornerRadius = try c.decodeIfPresent([String: Int].self, forKey: .cornerRadius) ?? [:]
    }
}

// MARK: - Spacing Configuration

public struct SpacingConfig: Codable {
    public var small: Int
    public var `default`: Int
    public var medium: Int
    public var large: Int
    public var extraLarge: Int
    public var padding: Int

    public init(
        small: Int = 4,
        default: Int = 8,
        medium: Int = 12,
        large: Int = 16,
        extraLarge: Int = 24,
        padding: Int = 16
    ) {
        self.small = small
        self.default = `default`
        self.medium = medium
        self.large = large
        self.extraLarge = extraLarge
        self.padding = padding
    }
}

// MARK: - Separator Configuration

public struct SeparatorConfig: Codable {
    public var lineThickness: Int
    public var lineColor: String

    public init(lineThickness: Int = 1, lineColor: String = "#E0E0E0") {
        self.lineThickness = lineThickness
        self.lineColor = lineColor
    }
}

// MARK: - Font Sizes Configuration

public struct FontSizesConfig: Codable {
    public var small: Int
    public var `default`: Int
    public var medium: Int
    public var large: Int
    public var extraLarge: Int

    public init(
        small: Int = 12,
        default: Int = 14,
        medium: Int = 17,
        large: Int = 21,
        extraLarge: Int = 26
    ) {
        self.small = small
        self.default = `default`
        self.medium = medium
        self.large = large
        self.extraLarge = extraLarge
    }
}

// MARK: - Font Weights Configuration

public struct FontWeightsConfig: Codable {
    public var lighter: Int
    public var `default`: Int
    public var bolder: Int

    public init(lighter: Int = 300, default: Int = 400, bolder: Int = 600) {
        self.lighter = lighter
        self.default = `default`
        self.bolder = bolder
    }
}

// MARK: - Font Types Configuration

public struct FontTypesConfig: Codable {
    public var `default`: FontTypeDefinition
    public var monospace: FontTypeDefinition

    public init(
        default: FontTypeDefinition = FontTypeDefinition(fontFamily: "System"),
        monospace: FontTypeDefinition = FontTypeDefinition(fontFamily: "Courier")
    ) {
        self.default = `default`
        self.monospace = monospace
    }
}

/// Full font type definition with optional per-type size/weight overrides
public struct FontTypeDefinition: Codable {
    public var fontFamily: String
    public var fontSizes: FontSizesConfig?
    public var fontWeights: FontWeightsConfig?

    public init(
        fontFamily: String,
        fontSizes: FontSizesConfig? = nil,
        fontWeights: FontWeightsConfig? = nil
    ) {
        self.fontFamily = fontFamily
        self.fontSizes = fontSizes
        self.fontWeights = fontWeights
    }
}

// MARK: - Container Styles Configuration

public struct ContainerStylesConfig: Codable {
    public var `default`: ContainerStyleConfig
    public var emphasis: ContainerStyleConfig
    public var good: ContainerStyleConfig
    public var attention: ContainerStyleConfig
    public var warning: ContainerStyleConfig
    public var accent: ContainerStyleConfig

    public init(
        default: ContainerStyleConfig = ContainerStyleConfig(
            backgroundColor: "#FFFFFF",
            foregroundColors: ForegroundColorsConfig()
        ),
        emphasis: ContainerStyleConfig = ContainerStyleConfig(
            backgroundColor: "#F5F5F5",
            foregroundColors: ForegroundColorsConfig()
        ),
        good: ContainerStyleConfig = ContainerStyleConfig(
            backgroundColor: "#E8F5E9",
            foregroundColors: ForegroundColorsConfig()
        ),
        attention: ContainerStyleConfig = ContainerStyleConfig(
            backgroundColor: "#FFF3E0",
            foregroundColors: ForegroundColorsConfig()
        ),
        warning: ContainerStyleConfig = ContainerStyleConfig(
            backgroundColor: "#FFEBEE",
            foregroundColors: ForegroundColorsConfig()
        ),
        accent: ContainerStyleConfig = ContainerStyleConfig(
            backgroundColor: "#E3F2FD",
            foregroundColors: ForegroundColorsConfig()
        )
    ) {
        self.default = `default`
        self.emphasis = emphasis
        self.good = good
        self.attention = attention
        self.warning = warning
        self.accent = accent
    }
}

public struct ContainerStyleConfig: Codable {
    public var backgroundColor: String
    public var borderColor: String
    public var foregroundColors: ForegroundColorsConfig

    public init(
        backgroundColor: String,
        foregroundColors: ForegroundColorsConfig,
        borderColor: String = "#E0E0E0"
    ) {
        self.backgroundColor = backgroundColor
        self.borderColor = borderColor
        self.foregroundColors = foregroundColors
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        backgroundColor = try c.decode(String.self, forKey: .backgroundColor)
        borderColor = try c.decodeIfPresent(String.self, forKey: .borderColor) ?? "#E0E0E0"
        foregroundColors = try c.decodeIfPresent(ForegroundColorsConfig.self, forKey: .foregroundColors) ?? ForegroundColorsConfig()
    }
}

public struct ForegroundColorsConfig: Codable {
    public var `default`: ColorConfig
    public var dark: ColorConfig
    public var light: ColorConfig
    public var accent: ColorConfig
    public var good: ColorConfig
    public var warning: ColorConfig
    public var attention: ColorConfig

    public init(
        default: ColorConfig = ColorConfig(default: "#000000", subtle: "#666666"),
        dark: ColorConfig = ColorConfig(default: "#000000", subtle: "#666666"),
        light: ColorConfig = ColorConfig(default: "#FFFFFF", subtle: "#CCCCCC"),
        accent: ColorConfig = ColorConfig(default: "#0078D4", subtle: "#0063B1"),
        good: ColorConfig = ColorConfig(default: "#4CAF50", subtle: "#388E3C"),
        warning: ColorConfig = ColorConfig(default: "#FF9800", subtle: "#F57C00"),
        attention: ColorConfig = ColorConfig(default: "#F44336", subtle: "#D32F2F")
    ) {
        self.default = `default`
        self.dark = dark
        self.light = light
        self.accent = accent
        self.good = good
        self.warning = warning
        self.attention = attention
    }
}

public struct ColorConfig: Codable {
    public var `default`: String
    public var subtle: String
    public var highlightColors: HighlightColorConfig

    public init(default: String, subtle: String, highlightColors: HighlightColorConfig = HighlightColorConfig()) {
        self.default = `default`
        self.subtle = subtle
        self.highlightColors = highlightColors
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.default = try c.decode(String.self, forKey: .default)
        subtle = try c.decode(String.self, forKey: .subtle)
        highlightColors = try c.decodeIfPresent(HighlightColorConfig.self, forKey: .highlightColors) ?? HighlightColorConfig()
    }
}

/// Highlight/selection colors for text within a color slot
public struct HighlightColorConfig: Codable {
    public var `default`: String
    public var subtle: String

    public init(default: String = "#FFFFFF00", subtle: String = "#FFFFFFE0") {
        self.default = `default`
        self.subtle = subtle
    }
}

// MARK: - Image Sizes Configuration

public struct ImageSizesConfig: Codable {
    public var small: Int
    public var medium: Int
    public var large: Int

    public init(small: Int = 60, medium: Int = 120, large: Int = 180) {
        self.small = small
        self.medium = medium
        self.large = large
    }
}

// MARK: - Actions Configuration

public struct ActionsConfig: Codable {
    public var actionsOrientation: String
    public var actionAlignment: String
    public var buttonSpacing: Int
    public var maxActions: Int
    public var spacing: String
    public var showCard: ShowCardConfig
    public var iconPlacement: String
    public var iconSize: Int

    public init(
        actionsOrientation: String = "Horizontal",
        actionAlignment: String = "Left",
        buttonSpacing: Int = 8,
        maxActions: Int = 5,
        spacing: String = "Default",
        showCard: ShowCardConfig = ShowCardConfig(),
        iconPlacement: String = "AboveTitle",
        iconSize: Int = 16
    ) {
        self.actionsOrientation = actionsOrientation
        self.actionAlignment = actionAlignment
        self.buttonSpacing = buttonSpacing
        self.maxActions = maxActions
        self.spacing = spacing
        self.showCard = showCard
        self.iconPlacement = iconPlacement
        self.iconSize = iconSize
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        actionsOrientation = try c.decodeIfPresent(String.self, forKey: .actionsOrientation) ?? "Horizontal"
        actionAlignment = try c.decodeIfPresent(String.self, forKey: .actionAlignment) ?? "Left"
        buttonSpacing = try c.decodeIfPresent(Int.self, forKey: .buttonSpacing) ?? 8
        maxActions = try c.decodeIfPresent(Int.self, forKey: .maxActions) ?? 5
        spacing = try c.decodeIfPresent(String.self, forKey: .spacing) ?? "Default"
        showCard = try c.decodeIfPresent(ShowCardConfig.self, forKey: .showCard) ?? ShowCardConfig()
        iconPlacement = try c.decodeIfPresent(String.self, forKey: .iconPlacement) ?? "AboveTitle"
        iconSize = try c.decodeIfPresent(Int.self, forKey: .iconSize) ?? 16
    }
}

public struct ShowCardConfig: Codable {
    public var actionMode: String
    public var style: String
    public var inlineTopMargin: Int

    public init(actionMode: String = "Inline", style: String = "Emphasis", inlineTopMargin: Int = 16) {
        self.actionMode = actionMode
        self.style = style
        self.inlineTopMargin = inlineTopMargin
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        actionMode = try c.decodeIfPresent(String.self, forKey: .actionMode) ?? "Inline"
        style = try c.decodeIfPresent(String.self, forKey: .style) ?? "Emphasis"
        inlineTopMargin = try c.decodeIfPresent(Int.self, forKey: .inlineTopMargin) ?? 16
    }
}

// MARK: - AdaptiveCard Configuration

public struct AdaptiveCardConfig: Codable {
    public var allowCustomStyle: Bool

    public init(allowCustomStyle: Bool = true) {
        self.allowCustomStyle = allowCustomStyle
    }
}

// MARK: - ImageSet Configuration

public struct ImageSetConfig: Codable {
    public var imageSize: String
    public var maxImageHeight: Int

    public init(imageSize: String = "Medium", maxImageHeight: Int = 100) {
        self.imageSize = imageSize
        self.maxImageHeight = maxImageHeight
    }
}

// MARK: - FactSet Configuration

public struct FactSetConfig: Codable {
    public var title: FactSetTextConfig
    public var value: FactSetTextConfig
    public var spacing: Int

    public init(
        title: FactSetTextConfig = FactSetTextConfig(weight: "Bolder"),
        value: FactSetTextConfig = FactSetTextConfig(weight: "Default"),
        spacing: Int = 8
    ) {
        self.title = title
        self.value = value
        self.spacing = spacing
    }
}

public struct FactSetTextConfig: Codable {
    public var weight: String
    public var size: String
    public var color: String
    public var isSubtle: Bool
    public var fontType: String
    public var wrap: Bool
    public var maxWidth: Int

    public init(
        weight: String = "Default",
        size: String = "Default",
        color: String = "Default",
        isSubtle: Bool = false,
        fontType: String = "Default",
        wrap: Bool = true,
        maxWidth: Int = 0
    ) {
        self.weight = weight
        self.size = size
        self.color = color
        self.isSubtle = isSubtle
        self.fontType = fontType
        self.wrap = wrap
        self.maxWidth = maxWidth
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        weight = try c.decodeIfPresent(String.self, forKey: .weight) ?? "Default"
        size = try c.decodeIfPresent(String.self, forKey: .size) ?? "Default"
        color = try c.decodeIfPresent(String.self, forKey: .color) ?? "Default"
        isSubtle = try c.decodeIfPresent(Bool.self, forKey: .isSubtle) ?? false
        fontType = try c.decodeIfPresent(String.self, forKey: .fontType) ?? "Default"
        wrap = try c.decodeIfPresent(Bool.self, forKey: .wrap) ?? true
        maxWidth = try c.decodeIfPresent(Int.self, forKey: .maxWidth) ?? 0
    }
}

// MARK: - Media Configuration

public struct MediaConfig: Codable {
    public var defaultPoster: String
    public var playButton: String
    public var allowInlinePlayback: Bool

    public init(
        defaultPoster: String = "",
        playButton: String = "",
        allowInlinePlayback: Bool = true
    ) {
        self.defaultPoster = defaultPoster
        self.playButton = playButton
        self.allowInlinePlayback = allowInlinePlayback
    }
}

// MARK: - Inputs Configuration

public struct InputsConfig: Codable {
    public var label: InputLabelGroupConfig
    public var errorMessage: ErrorMessageConfig

    public init(
        label: InputLabelGroupConfig = InputLabelGroupConfig(),
        errorMessage: ErrorMessageConfig = ErrorMessageConfig()
    ) {
        self.label = label
        self.errorMessage = errorMessage
    }
}

/// Label configuration for input groups (required vs optional)
public struct InputLabelGroupConfig: Codable {
    public var inputSpacing: String
    public var requiredInputs: InputLabelConfig
    public var optionalInputs: InputLabelConfig

    public init(
        inputSpacing: String = "Default",
        requiredInputs: InputLabelConfig = InputLabelConfig(),
        optionalInputs: InputLabelConfig = InputLabelConfig()
    ) {
        self.inputSpacing = inputSpacing
        self.requiredInputs = requiredInputs
        self.optionalInputs = optionalInputs
    }
}

/// Label appearance for a single input field
public struct InputLabelConfig: Codable {
    public var color: String
    public var isSubtle: Bool
    public var size: String
    public var suffix: String
    public var weight: String

    public init(
        color: String = "Default",
        isSubtle: Bool = false,
        size: String = "Default",
        suffix: String = "",
        weight: String = "Default"
    ) {
        self.color = color
        self.isSubtle = isSubtle
        self.size = size
        self.suffix = suffix
        self.weight = weight
    }
}

/// Error message styling for input validation
public struct ErrorMessageConfig: Codable {
    public var size: String
    public var spacing: String
    public var weight: String

    public init(
        size: String = "Default",
        spacing: String = "Default",
        weight: String = "Default"
    ) {
        self.size = size
        self.spacing = spacing
        self.weight = weight
    }
}

// MARK: - Host Width Configuration (Responsive Breakpoints)

public struct HostWidthConfig: Codable {
    public var veryNarrow: Int
    public var narrow: Int
    public var standard: Int

    public init(veryNarrow: Int = 0, narrow: Int = 0, standard: Int = 0) {
        self.veryNarrow = veryNarrow
        self.narrow = narrow
        self.standard = standard
    }
}

// MARK: - TextBlock Configuration

public struct TextBlockConfig: Codable {
    public var headingLevel: Int

    public init(headingLevel: Int = 2) {
        self.headingLevel = headingLevel
    }
}

// MARK: - Text Styles Configuration

public struct TextStylesConfig: Codable {
    public var heading: TextStyleConfig
    public var columnHeader: TextStyleConfig

    public init(
        heading: TextStyleConfig = TextStyleConfig(),
        columnHeader: TextStyleConfig = TextStyleConfig(weight: "Bolder")
    ) {
        self.heading = heading
        self.columnHeader = columnHeader
    }
}

/// Style definition for a semantic text role (heading, columnHeader, etc.)
public struct TextStyleConfig: Codable {
    public var weight: String
    public var size: String
    public var isSubtle: Bool
    public var color: String
    public var fontType: String

    public init(
        weight: String = "Default",
        size: String = "Default",
        isSubtle: Bool = false,
        color: String = "Default",
        fontType: String = "Default"
    ) {
        self.weight = weight
        self.size = size
        self.isSubtle = isSubtle
        self.color = color
        self.fontType = fontType
    }
}

// MARK: - Image Configuration

public struct ImageConfig: Codable {
    public var imageSize: String

    public init(imageSize: String = "Auto") {
        self.imageSize = imageSize
    }
}

// MARK: - Rating Configuration

public struct RatingElementConfig: Codable {
    public var filledStar: RatingStarConfig
    public var emptyStar: RatingStarConfig
    public var ratingTextColor: String
    public var countTextColor: String

    public init(
        filledStar: RatingStarConfig = RatingStarConfig(),
        emptyStar: RatingStarConfig = RatingStarConfig(),
        ratingTextColor: String = "#000000",
        countTextColor: String = "#000000"
    ) {
        self.filledStar = filledStar
        self.emptyStar = emptyStar
        self.ratingTextColor = ratingTextColor
        self.countTextColor = countTextColor
    }
}

public struct RatingStarConfig: Codable {
    public var marigoldColor: String
    public var neutralColor: String

    public init(
        marigoldColor: String = "#EAA300",
        neutralColor: String = "#212121"
    ) {
        self.marigoldColor = marigoldColor
        self.neutralColor = neutralColor
    }
}

// MARK: - Table Configuration

public struct TableConfig: Codable {
    public var cellSpacing: Int

    public init(cellSpacing: Int = 8) {
        self.cellSpacing = cellSpacing
    }
}

// MARK: - Compound Button Configuration

public struct CompoundButtonConfig: Codable {
    public var badge: BadgeConfig
    public var borderColor: String

    public init(
        badge: BadgeConfig = BadgeConfig(),
        borderColor: String = "#E1E1E1"
    ) {
        self.badge = badge
        self.borderColor = borderColor
    }
}

public struct BadgeConfig: Codable {
    public var backgroundColor: String

    public init(backgroundColor: String = "#5B5FC7") {
        self.backgroundColor = backgroundColor
    }
}
